File: adapters/rustsearch/grpc.go
package rustsearch

import (
	"context"
	"errors"

	"github.com/navidrome/navidrome/core/searchworker"
	"github.com/navidrome/navidrome/core/searchworker/gen"
)

func (e *Engine) startGRPC() error {
	client := searchworker.SearchClient()
	if client == nil {
		return searchworker.ErrNoGRPC
	}
	e.grpc = client
	return nil
}

func (e *Engine) grpcRoundTrip(ctx context.Context, req request) (response, error) {
	if e.grpc == nil {
		return response{}, errors.New("search gRPC client closed")
	}
	return searchworker.CallSearch(ctx, func(ctx context.Context, client gen.SearchClient) (response, error) {
		return e.grpcRoundTripOnce(ctx, client, req)
	})
}

func (e *Engine) grpcRoundTripOnce(ctx context.Context, client gen.SearchClient, req request) (response, error) {
	if client == nil {
		return response{}, errors.New("search gRPC client closed")
	}
	resp, err := client.Apply(ctx, toIndexRequest(req))
	if err != nil {
		return response{}, err
	}
	if resp.GetProtocol() != 0 && resp.GetProtocol() != protocolVersion {
		return response{}, errors.New("unsupported Rust search protocol")
	}
	if !resp.GetOk() {
		msg := resp.GetError()
		if msg == "" {
			msg = "Rust search request failed"
		}
		return response{}, errors.New(msg)
	}
	out := response{
		Protocol:   int(resp.GetProtocol()),
		OK:         true,
		Indexed:    resp.GetIndexed(),
		Normalized: resp.GetNormalized(),
	}
	if out.Protocol == 0 {
		out.Protocol = protocolVersion
	}
	for _, group := range resp.GetGroups() {
		hits := make([]hit, len(group.GetHits()))
		for i, h := range group.GetHits() {
			hits[i] = hit{ID: h.GetId(), Score: h.GetScore()}
		}
		out.Groups = append(out.Groups, searchGroup{Kind: group.GetKind(), Hits: hits})
	}
	return out, nil
}

func toIndexRequest(req request) *gen.IndexRequest {
	docs := toProtoDocuments(req.Documents)
	switch req.Op {
	case "begin_replace":
		return &gen.IndexRequest{Op: &gen.IndexRequest_BeginReplace{BeginReplace: &gen.BeginReplace{}}}
	case "append":
		return &gen.IndexRequest{Op: &gen.IndexRequest_Append{Append: &gen.Append{Documents: docs}}}
	case "commit_replace":
		return &gen.IndexRequest{Op: &gen.IndexRequest_CommitReplace{CommitReplace: &gen.CommitReplace{}}}
	case "abort_replace":
		return &gen.IndexRequest{Op: &gen.IndexRequest_AbortReplace{AbortReplace: &gen.AbortReplace{}}}
	case "upsert":
		return &gen.IndexRequest{Op: &gen.IndexRequest_Upsert{Upsert: &gen.Upsert{Documents: docs}}}
	case "delete":
		return &gen.IndexRequest{Op: &gen.IndexRequest_Delete{Delete: &gen.Delete{Keys: req.Keys}}}
	case "commit":
		return &gen.IndexRequest{Op: &gen.IndexRequest_Commit{Commit: &gen.Commit{}}}
	case "search_all":
		searches := make([]*gen.SearchSpec, len(req.Searches))
		for i, spec := range req.Searches {
			searches[i] = &gen.SearchSpec{Kind: spec.Kind, Offset: int32(spec.Offset), Limit: int32(spec.Limit)}
		}
		return &gen.IndexRequest{Op: &gen.IndexRequest_SearchAll{SearchAll: &gen.SearchAll{
			Query:      req.Query,
			LibraryIds: req.LibraryIDs,
			Searches:   searches,
		}}}
	case "normalize_fts":
		return &gen.IndexRequest{Op: &gen.IndexRequest_NormalizeFts{NormalizeFts: &gen.NormalizeFts{Values: req.Values}}}
	default:
		return &gen.IndexRequest{}
	}
}

func toProtoDocuments(docs []document) []*gen.Document {
	if len(docs) == 0 {
		return nil
	}
	out := make([]*gen.Document, len(docs))
	for i, doc := range docs {
		out[i] = &gen.Document{
			Key:        doc.Key,
			Id:         doc.ID,
			Kind:       doc.Kind,
			LibraryIds: doc.LibraryIDs,
			Primary:    doc.Primary,
			Secondary:  doc.Secondary,
		}
	}
	return out
}

File: rust/search/src/grpc.rs
use std::sync::{Arc, RwLock};

use tonic::{Request, Response, Status};

use crate::proto::search_server::Search;
use crate::proto::{
    HealthRequest, HealthResponse, Hit as ProtoHit, IndexRequest, IndexResponse,
    SearchGroup as ProtoGroup, index_request,
};
use crate::{Engine, Request as SearchOp, SearchDocument, SearchSpec, handle_request_locked};

pub struct SearchService {
    engine: Arc<RwLock<Option<Engine>>>,
}

impl SearchService {
    pub fn new() -> Self {
        Self {
            engine: Arc::new(RwLock::new(None)),
        }
    }
}

#[tonic::async_trait]
impl Search for SearchService {
    async fn apply(
        &self,
        request: Request<IndexRequest>,
    ) -> Result<Response<IndexResponse>, Status> {
        let engine = Arc::clone(&self.engine);
        let req = request.into_inner();
        let result = tokio::task::spawn_blocking(move || apply_sync(&engine, req))
            .await
            .map_err(|err| Status::internal(format!("search worker join: {err}")))?;
        Ok(Response::new(result))
    }

    async fn health(
        &self,
        _request: Request<HealthRequest>,
    ) -> Result<Response<HealthResponse>, Status> {
        Ok(Response::new(HealthResponse { ok: true }))
    }
}

fn apply_sync(engine: &RwLock<Option<Engine>>, request: IndexRequest) -> IndexResponse {
    let Some(op) = request.op else {
        return failed("op is required");
    };
    let search_op = match to_search_op(op) {
        Ok(op) => op,
        Err(error) => return failed(&error),
    };
    match handle_request_locked(engine, search_op) {
        Ok(response) => IndexResponse {
            protocol: response.protocol,
            ok: response.ok,
            groups: response
                .groups
                .into_iter()
                .map(|group| ProtoGroup {
                    kind: group.kind,
                    hits: group
                        .hits
                        .into_iter()
                        .map(|hit| ProtoHit {
                            id: hit.id,
                            score: hit.score,
                        })
                        .collect(),
                })
                .collect(),
            indexed: response.indexed,
            error: response.error.unwrap_or_default(),
            normalized: response.normalized.unwrap_or_default(),
        },
        Err(error) => failed(&format!("{error:#}")),
    }
}

fn to_search_op(op: index_request::Op) -> Result<SearchOp, String> {
    match op {
        index_request::Op::BeginReplace(_) => Ok(SearchOp::BeginReplace),
        index_request::Op::Append(append) => Ok(SearchOp::Append {
            documents: to_docs(append.documents),
        }),
        index_request::Op::CommitReplace(_) => Ok(SearchOp::CommitReplace),
        index_request::Op::AbortReplace(_) => Ok(SearchOp::AbortReplace),
        index_request::Op::Upsert(upsert) => Ok(SearchOp::Upsert {
            documents: to_docs(upsert.documents),
        }),
        index_request::Op::Delete(delete) => Ok(SearchOp::Delete { keys: delete.keys }),
        index_request::Op::Commit(_) => Ok(SearchOp::Commit),
        index_request::Op::SearchAll(search) => Ok(SearchOp::SearchAll {
            query: search.query,
            library_ids: search.library_ids,
            searches: search
                .searches
                .into_iter()
                .map(|spec| SearchSpec {
                    kind: spec.kind,
                    offset: spec.offset.max(0) as usize,
                    limit: spec.limit.max(0) as usize,
                })
                .collect(),
        }),
        index_request::Op::NormalizeFts(norm) => Ok(SearchOp::NormalizeFts { values: norm.values }),
    }
}

fn to_docs(docs: Vec<crate::proto::Document>) -> Vec<SearchDocument> {
    docs.into_iter()
        .map(|doc| SearchDocument {
            key: doc.key,
            id: doc.id,
            kind: doc.kind,
            library_ids: doc.library_ids,
            primary: doc.primary,
            secondary: doc.secondary,
        })
        .collect()
}

fn failed(error: &str) -> IndexResponse {
    IndexResponse {
        protocol: crate::PROTOCOL_VERSION,
        ok: false,
        error: error.to_string(),
        ..Default::default()
    }
}
